### PR TITLE
[`feat`] Add triplets/n-tuple support to AnglE

### PR DESCRIPTION
**Pull request by tomaarsen**
Hello!

## Pull Request overview
* Experiment to add triplets/n-tuple support to AnglE
* A training script based on all-nli 

## Details
This is a draft to add triplets/n-tuple support to AnglELoss, at which point the inputs are converted to pairs again, with labels set to 1 for positive pairs and 0 for negative pairs. I copied a simple training script and updated it to use AnglE, resulting in this model:
- https://huggingface.co/tomaarsen/distilroberta-base-nli-angle

Which can be fairly compared to e.g. these:
- https://huggingface.co/tomaarsen/distilroberta-base-nli-v2 (MultipleNegativesRankingLoss)
- https://huggingface.co/tomaarsen/distilroberta-base-nli-v3 (GISTEmbedLoss)

---

- Tom Aarsen

### CODE REVIEW

File: sentence_transformers/losses/AnglELoss.py [hunk 1]
```diff
@@ -68,6 +79,26 @@ def __init__(self, model: SentenceTransformer, scale: float = 20.0) -> None:
...
+        if len(embeddings) > 2:
+            # Anchor-positive-negative-... n-tuples case, convert to pairwise comparisons again, but with `labels` (or 1s)
+            # for the positive pairs and `1 - labels` for the negatives.
+            anchor_embeddings = embeddings[0]
+            combined_embeddings = [anchor_embeddings.repeat(len(embeddings) - 1, 1), torch.cat(embeddings[1:], dim=0)]
+
+            if labels is None:
+                labels = torch.ones(anchor_embeddings.size(0), device=anchor_embeddings.device)
+
+            combined_labels = torch.cat([labels, (1 - labels).repeat(len(embeddings) - 2)], dim=0)
```
[the remark this answers is not in the document]
Comment (tomaarsen) [reply sentence_transformers/losses/AnglELoss.py:94]: I'm fine with `labels` not being recommended, but technically possible to use.
